### ISSUE TITLE
Ignore .gitignored files

### ISSUE THREAD

**Issue by robatwilliams** [facebook/watchman#904]
I'm configuring Watchman for use with Jest. Everything I add to `ignore_dirs` (build output, node_modules) is already in my `.gitignore`. I'd like to be able to configure watchman to refer to that ignore file, instead of repeating it.

**Comment by wez** [comment]
While this sounds like a nice convenience, it's fraught with difficulty.  Watchman is used by both mercurial and git to provide lists of files when considering the state of the working copy.

Paths that are ignored by watchman are effectively permanently invisible to the `status` function in source control.  That means that it becomes impossible to correctly handle the case where someone explicitly tracks and adds files that match gitignore rules--that's an anti-pattern in general, but happens much more often than anyone would like.

Another complication is that correct gitignore handling is the result of composing ignore rules from layered/cascading ignore files within the repo, above the repo, from the home directory and from a system default configuration file.  That makes it difficult to reason about the complexity of ignore processing, because there is potentially a non-trivial amount of IO processing just to discover the set of ignore files.

Related to that: ignore files can change at any time; either through user action or for example by checking out across a commit where the ignore files were changed.  Since watchman doesn't watch directories that match the ignore rules (that's the purpose of watchman ignores, after all), correctly processing an ignore change would mean re-scanning the repository to re-assess and re-watch locations that were previously ignored.  That means that the IO costs around something that should be trivial checkout operation can increase in undesirable and disproportionately large ways in very large repos.  There are pathological situations around `rebase` and `bisect` when a series of commits change the ignore files also.  This is also the reason why changes to `.watchmanconfig` are not automatically processed by watchman.

The best practice recommendations are:

* Where possible, place build outputs outside the repo.  Even with watchman configured to ignore the directories, on macOS the kernel only supports ignoring a limited number of paths (8!), which means that the system is still delivering notifications for some of those paths and that puts pressure on the system as a whole.  Watchman has logic to filter those out of the results, but your system paid the cost of sending the notifications to watchman anyway.
* Only specify paths that will never be checked into a repo in the `.watchmanconfig` file

**Comment by gpshead** [comment]
I think this was closed before github differentiated between Completed and Not Planned.  Based on @wez's comment and lack of PR or commit refs, I presume @robatwilliams closing this was more of a "Not Planned" issue close?

**Comment by wez** [comment]
Not Planned. Will almost certainly never be implemented.

**Comment by gpshead** [comment]
To leave breadcrumbs for future selves: I wound up here while researching because of some code we've got translating from gitignore glob syntax into watchman match glob syntax and getting that pedantically correct has proven... painful.  _(Insert standard meme quote about using regexes and problems increasing here)_

**Comment by robatwilliams** [comment]
> I presume @robatwilliams closing this was more of a "Not Planned" issue close?

Agreed, I must have used the default close button and not dropped down the arrow to select not planned

**Comment by omry** [comment]
Created a [tiny tool](https://github.com/omry/watchmanignore) that generates a conservative .watchmanconfig from .gitignore files.
